### ISSUE TITLE
[jest/expect-expect] doesn't work with assertion function names with $ character

### ISSUE THREAD

**Issue by nsasinovich** [jest-community/eslint-plugin-jest#626]
# Problem
I have configured `jest/expect-expect` the next way:
```
"jest/expect-expect": ["warn", { "assertFunctionNames": ["expect", "expect$"] }]
```
But as I explored the rule doesn't work with the function names which contain special characters from regular expressions, including `$`.

As I see from [the implementation of matching function](https://github.com/jest-community/eslint-plugin-jest/blob/d5ab9b46e4fcccd2c7b395ed7b5439a69426eecd/src/rules/expect-expect.ts#L24-L41) provided in `assertFunctionNames` patterns are transformed into regular expressions for further matching. Thus, characters as `$` are handled as regex special characters, not like part of a provided function name.

# Expectation
I expect that assertion function names with special characters are whitelisted and do not trigger eslint errors.

# Proposed solution
As a solution, I propose to add an extra step and escape special characters while converting provided function name into regular expression. 
``` js
function matchesAssertFunctionName(
  nodeName: string,
  patterns: readonly string[],
): boolean {
  return patterns.some(p =>
    new RegExp(
      `^${p
        .split('.')
        .replace((/[+?^${}()|[\]\\]/g, '\\$&'))
        .map(x => {
          if (x === '**') return '[a-z\\.]*';

          return x.replace(/\*/gu, '[a-z]*');
        })
        .join('\\.')}(\\.|$)`,
      'ui',
    ).test(nodeName),
  );
}
```

NOTE: [the link to syntax characters list](https://tc39.es/ecma262/#prod-SyntaxCharacter). I've excluded `*`, `.` as they could be used in wildcards syntax of names.

**Comment by G-Rath** [comment]
This is naturally supported already, you just need to escape the character in question using `\\`:

```
'jest/expect-expect': [
  'warn',
  { assertFunctionNames: ['expect', 'expect\\$'] }
]
```

**Comment by nsasinovich** [comment]
Thanks for fast feedback!

It's also an option but then it probably should be made clearer in the rule documentation.

**Comment by G-Rath** [comment]
No problem :)

You're totally right that it'll be good to have an example for this - I've done so via #627, and also added a test to ensure we support this situation.

**Comment by nsasinovich** [comment]
Thank you!
